ergo-chain-sync/src/client/types.rs

```diff
@@ -1,30 +1,48 @@
-use std::fmt::{Display, Formatter};
+use std::{
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
 
-use serde::{Deserialize, Serialize};
+use isahc::http::{uri::InvalidUri, Uri};
 
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct Url(String);
+#[derive(Debug, Clone)]
+pub struct Url(Uri);
 
 impl Display for Url {
     fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
-        f.write_str(&self.0)
+        self.0.fmt(f)
     }
 }
 
-impl From<String> for Url {
-    fn from(value: String) -> Self {
-        Self(value)
-    }
+#[derive(Debug)]
+pub enum InvalidUrl {
+    Uri(InvalidUri),
+    NoScheme,
 }
 
-impl From<&str> for Url {
-    fn from(value: &str) -> Self {
-        Self(String::from(value))
+impl FromStr for Url {
+    type Err = InvalidUrl;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let uri = Uri::from_str(s).map_err(InvalidUrl::Uri)?;
+        if uri.scheme_str().is_none() {
```

Comment:
We can create a `Uri` without the scheme (http etc.) but I think it's good to make it mandatory.